goodwe/et.py

```diff
@@ -459,12 +521,99 @@ async def read_device_info(self):
+        # if not is_4_mptt(self):
+        #     # This inverter does not have 4th MPPTs
+        #     self._sensors = tuple(filter(lambda s: not ('pv4' in s.id_), self._sensors))
+        #     if not is_3_mptt(self):
+        #         # This inverter neither has 3rd MPPTs
+        #         self._sensors = tuple(filter(lambda s: not ('pv3' in s.id_), self._sensors))
+        #
+        # wrong: only ppv4 / ppv3 have to be filtered, ipv and vpv follow other rules
+        #
+        # solution:
+        #  ppv* and *mppt are related to number of MPPT's
+        #  ipv's and vpv's are related to the number of strings of PV-modules
+        #
+        # filter  self._sensor and self._sensors_mptt
+        #   filter ipv# and vpv# according to number of strings of model
+        #   filter imppt# and pmppt# according to number of mppt's
+        #
+        # ADDED: gowima
+        # list based filters for _all_sensors and _all_mppt_sensors
+        # see also is_2_strings_per_mppt() in model.py
+        #
+        # COMMENT: Variables and functions are named "mPTt" but labels of sensors "mPPt".
+        #          It should be mppt (Maximum Power Point Tracking)
+
+        # sensor.id_ as filter criteria
+        mppt2_sensor_f = ["ppv3", "pv3_mode", "pv3_mode_label",
+                          "ppv4", "pv4_mode", "pv4_mode_label"]
+        mppt3_sensor_f = ["ppv4", "pv4_mode", "pv4_mode_label"]
+        mppt4_sensor_f = []
+
+        # sensor.id_ "ends with" as filter criteria
+        # for 2 MPPTs with 1 string per MPPT
+        mppt2_1string_f = ["pv3", "pv4", "pv5", "pv6", "pv7", "pv8", "pv9", "pv10",
+                           "pv11", "pv12", "pv13", "pv14", "pv15", "pv16",
+                           "mppt3", "mppt4", "mppt5", "mppt6", "mppt7", "mppt8"]
+        # for 2 MPPTs with 2 strings per MPPT
+        mppt2_2string_f = ["pv5", "pv6", "pv7", "pv8", "pv9", "pv10",
+                           "pv11", "pv12", "pv13", "pv14", "pv15", "pv16",
+                           "mppt3", "mppt4", "mppt5", "mppt6", "mppt7", "mppt8"]
+        # for 3 MPPTs with 1 string per MPPT
+        mppt3_1string_f = ["pv4", "pv5", "pv6", "pv7", "pv8", "pv9", "pv10",
+                           "pv11", "pv12", "pv13", "pv14", "pv15", "pv16",
+                           "mppt4", "mppt5", "mppt6", "mppt7", "mppt8"]
+        # for 3 MPPTs with 2 strings per MPPT
+        mppt3_2string_f = ["pv7", "pv8", "pv9", "pv10",
+                           "pv11", "pv12", "pv13", "pv14", "pv15", "pv16",
+                           "mppt4", "mppt5", "mppt6", "mppt7", "mppt8"]
+        # for 4 MPPTs with 2 strings per MPPT
+        mppt4_2string_f = ["pv9", "pv10", "pv11", "pv12", "pv13", "pv14", "pv15", "pv16",
+                           "mppt5", "mppt6", "mppt7", "mppt8"]
+
+        # set filter lists based on parameters of inverter model
+
+        # implicitly 4 mppts come with 2 strings per mppt
+        # TODO: CHECK if device with 4mppt and 1 string per mppt exists
```

Comment:
It seems it does exist - GE's white label GEH model.
https://www.gesolarinverter.com/images/FILES/DOWNLOADS/1-GLOBAL/1-DATASHEETS/GEH%205-10kW_G.pdf